src/analysis/von_neumann_algebra/basic.lean

```diff
@@ -0,0 +1,96 @@
+
+So that we can unambiguously talk about these "abstract" von Neumann algebras
+in parallel with the "concrete" ones (weakly closed *-subalgebras of B(H)),
+we name this definition `wstar_algebra`.
+-/
+class wstar_algebra (M : Type u) [normed_ring M] [star_ring M] [cstar_ring M]
+  [module ℂ M] [normed_algebra ℂ M] [star_module ℂ M] :=
+(exists_predual : ∃ (X : Type u) [normed_group X] [normed_space ℂ X] [complete_space X],
+  nonempty (normed_space.dual ℂ X ≃ₗᵢ⋆[ℂ] M))
+
+open_locale inner_product
+
+/-- The commutant of a subsemiring. -/
+def subsemiring.commutant {A : Type u} [semiring A] (B : subsemiring A) : subsemiring A :=
+{ carrier := { x : A | ∀ y ∈ B, x * y = y * x },
+  zero_mem' := by { intros y h, simp, },
+  one_mem' := by { intros y h, simp, },
+  mul_mem' := λ a b ha hb c mc, by rw [mul_assoc, hb _ mc, ←mul_assoc, ha _ mc, mul_assoc],
+  add_mem' := λ a b ha hb c mc, by rw [add_mul, mul_add, ha _ mc, hb _ mc], }
+
+/-- The commutant of a subalgebra. -/
+def subalgebra.commutant {𝕜 : Type v} [comm_semiring 𝕜] {A : Type u} [semiring A] [algebra 𝕜 A]
+  (B : subalgebra 𝕜 A) : subalgebra 𝕜 A :=
+{ algebra_map_mem' := λ r x mx, by rw [algebra.commutes],
+  .. B.to_subsemiring.commutant, }
+
+set_option old_structure_cmd true
+
+/-- A *-subalgebra is a sub
+algebra of a *-algebra which is closed under *. -/
+structure star_subalgebra (R : Type u) (A : Type v) [comm_semiring R] [star_ring R]
+  [semiring A] [star_ring A] [algebra R A] [star_module R A] extends subalgebra R A : Type v :=
+(star_mem' {a} : a ∈ carrier → star a ∈ carrier)
+
+namespace star_subalgebra
+variables (R : Type u) (A : Type v) [comm_semiring R] [star_ring R]
+  [semiring A] [star_ring A] [algebra R A] [star_module R A]
+
+instance : set_like (star_subalgebra R A) A :=
+⟨star_subalgebra.carrier, λ p q h, by cases p; cases q; congr'⟩
+
+instance (R : Type u) (A : Type v) [comm_semiring R] [star_ring R]
+  [semiring A] [star_ring A] [algebra R A] [star_module R A] : inhabited (star_subalgebra R A) :=
+⟨{ star_mem' := by tidy, ..(⊤ : subalgebra R A) }⟩
+
+end star_subalgebra
+
+/-- The commutant of a *-subalgebra. -/
+def star_subalgebra.commutant {R : Type u} {A : Type v} [comm_semiring R] [star_ring R]
+  [semiring A] [star_ring A] [algebra R A] [star_module R A] (B : star_subalgebra R A) :
+  star_subalgebra R A :=
+{ star_mem' := λ x xm y hy, by simpa using congr_arg star (xm _ (B.star_mem' hy)).symm,
+  ..B.to_subalgebra.commutant, }
+
+/--
+The double commutant definition of a von Neumann algebra,
+as a *-closed subalgebra of bounded operators on a Hilbert space,
+which is equal to its double commutant.
+
+Note that this definition is parameterised by the Hilbert space
+on which the algebra faithfully acts.
+-/
+@[nolint has_inhabited_instance]
+structure von_neumann_algebra (H : Type u) [inner_product_space ℂ H] [complete_space H] extends
+  M : star_subalgebra ℂ (H →L[ℂ] H) :=
+(double_commutant : M.commutant.commutant = M)
```

Comment:
```suggestion
@[nolint has_inhabited_instance]
structure von_neumann_subalgebra (H : Type u) [inner_product_space ℂ H] [complete_space H] extends
  M : star_subalgebra ℂ (H →L[ℂ] H) :=
(double_commutant : M.commutant.commutant = M)
```
I think calling this a "subalgebra" would be better, otherwise it's a bit strange that `H →L[ℂ] H` itself is not a von Neumann algebra.

Reply:
I'm inclined to agree that subalgebra would be better, even if non-standard. The only hesitation would be: how often will we want sub-subalgebras? My gut feeling is "not frequently," because I think we will more often work in the abstract w-star setting and have subalgebras of w-star algebras.

Reply:
I'm confused, why would `H →L[ℂ] H` not be a von Neumann algebra? In its `star_subalgebra` guise it is just `\top`.

My instinct is that we're going to want this "concrete" version quite a lot, even if we have `wstar_algebra` available as well. People in the von Neumann algebra world are used to carrying around a Hilbert space! I think the fact that I'm proposing *implementing* this via `star_subalgebra` should just be considered an implementation detail. This is just a von Neumann algebra equipped with a favourite faithful representation, which is exactly the usual "concrete" definition.

Reply:
> I'm confused, why would `H →L[ℂ] H` not be a von Neumann algebra? In its `star_subalgebra` guise it is just `\top`.

Right, but this enough to mean that you can't directly use a lemma about von Neumann (sub)algebras on `H →L[ℂ] H`.

> My instinct is that we're going to want this "concrete" version quite a lot, even if we have `wstar_algebra` available as well. People in the von Neumann algebra world are used to carrying around a Hilbert space! I think the fact that I'm proposing _implementing_ this via `star_subalgebra` should just be considered an implementation detail. This is just a von Neumann algebra equipped with a favourite faithful representation, which is exactly the usual "concrete" definition.

I agree this is going to be used a lot -- certainly it's the obvious thing to do in quantum information.

Reply:
> Right, but this is enough to mean that you can't directly use a lemma about von Neumann (sub)algebras on `H →L[ℂ] H`.

Sorry, I still don't understand what you mean.

Reply:
Maybe I'm missing something, but suppose we have, say, a theorem that says that an element of a von Neumann algebra has a spectral decomposition, how do we apply it to `a : E →L[ℂ] E` without having to converting to an element of `top`?

Reply:
>  We might even call them `von_neumann_algebra` and `von_neumann_subalgebra`.

This wasn't quite clear to me, but if you were arguing for renaming `wstar_algebra` to `von_neumann_algebra` then I would strongly disagree with that; it doesn't match the literature and then it would be confusing that `von_neumann_algebra` and `von_neumann_subalgebra` are very different beasts (and *especially* confusing that `von_neumann_subalgebra` is *not* a subalgebra of `von_neumann_algebra`)

Reply:
I don't think it's too weird that `E →L[ℂ] E` is not itself a von Neumann algebra without conversion to `top`. It will be a `wstar_algebra` naturally once we show it is the dual of the trace-class (of course we'll need to define the trace-class first!).

Reply:
> > We might even call them `von_neumann_algebra` and `von_neumann_subalgebra`.
> 
> This wasn't quite clear to me, but if you were arguing for renaming `wstar_algebra` to `von_neumann_algebra` then I would strongly disagree with that; it doesn't match the literature and then it would be confusing that `von_neumann_algebra` and `von_neumann_subalgebra` are very different beasts (and _especially_ confusing that `von_neumann_subalgebra` is _not_ a subalgebra of `von_neumann_algebra`)

I see -- I actually thought W*-algebras and von Neumann algebras were basically synonyms. In that case `wstar_algebra` and `von_neumann_subalgebra` sounds good.

Reply:
> I don't think it's too weird that `E →L[ℂ] E` is not itself a von Neumann algebra without conversion to `top`. It will be a `wstar_algebra` naturally once we show it is the dual of the trace-class (of course we'll need to define the trace-class first!).

Right -- I'm happy with the two definitions as stated, all I'm arguing for is to have a name that makes it clear that one of them is a subobject.

Reply:
I'm still not really convinced. I mean, someone who works with von Neumann algebras I think would say "duh" if you pointed out that your von Neumann algebra acting on H was a subalgebra of B(H). And eventually we are going to want to talk about subalgebras (e.g. subfactors!) and I don't think we'd be happy about writing `subsubalgebra`.

What about `von_neumann_algebra_on H` or `von_neumann_algebra_acting_on H`? My inclination is that the extra characters are not worth the gain in clarity, but I'd prefer these to `von_neumann_subalgebra H`.

Reply:
I think I'd rather have `von_neumann_algebra` than either of those actually! Just out of curiosity, to get `A` to be a subalgebra of `B`, do we need to do more than just have `A B : von_neumann_algebra H` with `A ≤ B`?